libpysal/weights/util.py

```diff
@@ -211,6 +207,82 @@ def lat2W(nrows=5, ncols=5, rook=True, id_type='int', **kwargs):
     return W(w, weights, ids=ids, id_order=ids[:], **kwargs)
 
 
+def xrast2W(rasterf, rook=True, **kwargs):
+    """
+    Create a W object from rasters(xarray.DataArray)
+    
+    Parameters
+    ----------
+    raster     : xarray.DataArray
+                 raster file accessed using xarray.open_rasterio method
+    rook       : boolean
+                 type of contiguity. Default is rook. For queen, rook =False
+    **kwargs   : keyword arguments
+                 optional arguments for :class:`pysal.weights.W`
+    Returns
+    -------
+    w    : libpysal.weights.W
+           instance of spatial weights class W
+    data : pandas.Series
+           Values from `raster` as a vector of dimension (`w.n` x
+           None) aligned with `w`            
+    """
+    try:
+        import pandas as pd
+        import xarray as xr
+    except ImportError:
+        raise ImportError(
+            "pandas and xarray must be installed to use this method")
+
+    w = lat2W(*rasterf[0].shape, rook=rook, **kwargs)
+    if isinstance(rasterf, xr.DataArray):
+        rasterf_masked = rasterf.where(rasterf.data != rasterf.nodatavals[0])
+        rasterf_masked = rasterf_masked[0].data.flatten()
+        id_order = np.argwhere(~np.isnan(rasterf_masked)).flatten()
+        w = w_subset(w, id_order)
+        data = rasterf_masked[~np.isnan(rasterf_masked)]
+        data = pd.Series(data, index=id_order)
+    return w, data
+
+
+def xrast2SW(rasterf, criterion="rook"):
+    """
+    Generate a sparse W matrix from rasters(xarray.DataArray)
+    
+    Parameters
+    ----------
+    raster     : xarray.DataArray
+                 raster file accessed using xarray.open_rasterio method
+    rook       : {"rook", "queen"}
```

Comment:
> your parameters do not match up here. Generally I think it's a better idea to not allow for` rook=True`, but to give an option as seen here, that allows us to add other functionality later on.

This is because one function followed [`lat2W`](https://github.com/pysal/libpysal/blob/16a7f585bb95012bdd3dc9267bf48065f2afd1fa/libpysal/weights/util.py#L128) and other followed [`lat2SW`](https://github.com/pysal/libpysal/blob/16a7f585bb95012bdd3dc9267bf48065f2afd1fa/libpysal/weights/util.py#L1126). 
Now that we are using `contiguity.from_xarray`, so contiguity argument is not needed anymore... will try to keep the structure consistent :)